File: js/node/testing/node-test-notes/lib/calculator.js
export default class Calculator {
  // ---
  // Basic
  // ---

  /** Add two numbers */
  add(a, b) {
    return a + b;
  }
  /** Subtract two numbers */
  subtract(a, b) {
    return a - b;
  }
  /** Multiply two numbers */
  multiply(a, b) {
    return a * b;
  }
  /** Divide two numbers */
  divide(a, b) {
    return a / b;
  }

  // ---
  // Aggregations
  // ---

  /** Get the sum of the provided numbers */
  getSum(numbers) {
    return numbers.reduce((acc, curr) => acc + curr, 0);
  }
  /** Get the product of the provided numbers */
  getProduct(numbers) {
    return numbers.reduce((acc, curr) => acc * curr, 1);
  }
  /** Get the mean of the provided numbers */
  getMean(numbers) {
    return numbers?.length ? this.getSum(numbers) / numbers.length : Number.NaN;
  }
  /** Get the standard deviation of the provided numbers */
  getStd(numbers, completeSample = false) {
    const n = numbers?.length;
    if (!n) {
      return Number.NaN;
    }
    const mean = this.getMean(numbers);
    const sumOfSquaredDiffs = numbers.reduce(
      (acc, curr) => acc + Math.pow(curr - mean, 2),
      0,
    );
    const populationSize = completeSample ? n : n - 1;
    const std = Math.sqrt(sumOfSquaredDiffs / populationSize);
    return std;
  }

  // ---
  // Random
  // ---

  /** Pretend to get a random integer asynchronously */
  async randInt(max) {
    await new Promise((r) => setTimeout(r, 10));
    return Math.ceil(Math.random() * max);
  }
}

File: go/testing/gomega-notes/lib/calculator/calculator.go
package calculator

import (
	"math"
)

type Number interface {
	int64 | int32 | int16 | int8 | int | float64 | float32 | uint64 | uint32 | uint16 | uint8
}

// ---
// Basic
// ---

// Add adds two numbers
func Add[T Number](a, b T) T {
	return a + b
}

// Subtract subtracts two numbers
func Subtract[T Number](a, b T) T {
	return a - b
}

// Multiply multiplies two numbers
func Multiply[T Number](a, b T) T {
	return a * b
}

// Divide divides two numbers
func Divide[T Number](a, b T) float64 {
	if b == 0 {
		return math.NaN()
	}
	return float64(a) / float64(b)
}

// ---
// Aggregations
// ---

func reduce[I any, O any](items []I, reducerFunc func(acc O, curr I) O, initial O) O {
	result := initial
	for _, item := range items {
		result = reducerFunc(result, item)
	}
	return result
}

// Sum gets the sum of a slice of numbers
func Sum[T Number](numbers []T) T {
	sum := reduce(
		numbers,
		func(acc, curr T) T {
			return acc + curr
		},
		0,
	)
	return sum
}

// Product gets the product of a slice of numbers
func Product[T Number](numbers []T) T {
	product := reduce(
		numbers,
		func(acc, curr T) T {
			return acc * curr
		},
		1,
	)
	return product
}

// Mean gets the mean of a slice of numbers
func Mean[T Number](numbers []T) float64 {
	length := len(numbers)
	if length == 0 {
		return math.NaN()
	}
	sum := Sum(numbers)
	return float64(sum) / float64(length)

}

// Std gets the standard deviation of a slice of numbers
func Std[T Number](numbers []T, completeSample bool) float64 {
	length := len(numbers)
	if length == 0 {
		return math.NaN()
	}
	mean := Mean(numbers)
	sumOfSquaredDiffs := reduce(
		numbers,
		func(acc float64, curr T) float64 {
			return acc + math.Pow(float64(curr)-mean, 2)
		},
		0.0,
	)

	populationSize := length
	if !completeSample {
		populationSize -= 1
	}
	std := math.Sqrt(sumOfSquaredDiffs / float64(populationSize))
	return std
}
